lib/transformer.js

```diff
@@ -0,0 +1,155 @@
+'use strict'
+
+const zlib = require('zlib')
+const aws = require('aws-sdk')
+const {
+  checkFileFormat
+, hasProperty
+} = require('./utils.js')
+
+const s3 = new aws.S3({
+  apiVersion: '2006-03-01'
+})
+
+module.exports = {
+  extractData
+, getLogs
+, prepareLogs
+, s3
+}
+
+function extractData(data) {
+  if (!data || !Object.keys(data).length) return []
+
+  const {
+    Records
+  , logFiles
+  , ...meta
+  } = data
+
+  const base = {
+    timestamp: Date.now()
+  }
+
+  let records = data
+  if (Records) {
+    records = Records
+    base.meta = meta
+  } else if (logFiles) {
+    records = logFiles
+    base.meta = meta
+  }
+
+  if (meta && meta.eventTime) {
+    base.timestamp = (new Date(meta.eventTime)).getTime()
+  }
+
+  if (!Array.isArray(records)) {
+    const lineJSON = {
+      line: records
+    , ...base
+    }
+
+    if (typeof records !== 'string') {
+      lineJSON.line = JSON.stringify(records)
+    }
+
+    return [lineJSON]
+  }
+
+  return records.map((record) => {
+    const lineJSON = {
+      line: record
+    , ...base
+    }
+
+    if (typeof record !== 'string') {
+      lineJSON.line = JSON.stringify(record)
+    }
+
+    if (record.eventTime) {
+      lineJSON.timestamp = (new Date(record.eventTime)).getTime()
+    }
+
+    return lineJSON
+  })
+}
+
+async function getLogs(params) {
+  let err
+  if (!hasProperty(params, 'Key') || !hasProperty(params, 'Bucket')) {
+    err = new Error('Both Bucket and Key params must be provided')
+    err.meta = {params}
+    throw err
+  }
+
+  let data
+  try {
+    data = await s3.getObject(params)
+  } catch (error) {
+    err = new Error(`Error in Getting ${params.Bucket}/${params.Key}: ${error}`)
+    err.meta = {error, params}
+    throw err
+  }
+
+  if (!hasProperty(data, 'Body')) {
+    err = new Error(`Corrupted data returned from ${params.Bucket}/${params.Key}`)
+    err.meta = {params}
+    throw err
+  }
+
+  let content = data.Body
+  const type = checkFileFormat(params.Key)
+  if (type.gz) {
+    try {
+      content = zlib.gunzipSync(content)
+    } catch (error) {
+      err = new Error(`Error in Unzipping ${params.Bucket}/${params.Key}: ${error}`)
```

Comment:
we don't need to include the `params` or the original `error` in the message, including it in `error.meta` is sufficient (here and below)

Reply:
Done!